Support client credentials flow in authorization

D-McAdams opened the pull request:
This PR adds support for OAuth 2.1 Client Credentials Flow to enable machine-to-machine authentication in the MCP authorization specification. See SEP https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1046

D-McAdams commented:
Hmm, the diffs are a bit messy. The reason for that is I had to uplevel the section on "Resource Parameter Implementation" out of the authorization code flow, in order to make it commonly applicable to both authorization_code and client_credentials flow. None of the content for that section was intended to be changed.

pcarleton commented on docs/specification/draft/basic/authorization.mdx:
```diff
@@ -24,6 +24,30 @@ Authorization is **OPTIONAL** for MCP implementations. When supported:
+
+### Authorization Flows
+
+This specification defines two OAuth 2.1 flows that implementations **SHOULD** support:
```
I'm wondering if this is a "profile" thing that we need to formalize. 

As in, the "maximum interoperability for human interactive MCP clients" would have a  SHOULD (minimum) for authorization code, and then a "machine to machine" profile would have a SHOULD on client credentials flow.

There's not a big benefit for the machine to machine profile to support authorization code, and similarly there's not a big benefit for a human-based flow to support machine-to-machine.  There's the case of someone wanting to implement both profiles, and that's fine.

aaronpk replied:
Yeah, similar with the enterprise-managed profile, which doesn't apply to MCP servers that only have consumer users and don't have users logging in through an enterprise IdP. I like the idea of profiles to make it explicit which parts of OAuth are needed to be supported by the AS.

D-McAdams replied:
I expect we may revisit the spec structure again as we finalize the plan for profiles/extensions. For now, want to ensure we unblock audiences building autonomous agents. The original version of the MCP Auth spec did mention client credentials and it was dropped in the subsequent edits. This restores it, alongside guidance about when to implement it.

LucaButBoring commented on docs/specification/draft/basic/authorization.mdx:
```diff
@@ -24,6 +24,30 @@ Authorization is **OPTIONAL** for MCP implementations. When supported:
+
+This specification defines two OAuth 2.1 flows that implementations **SHOULD** support:
+
+1. Authorization Code Flow - Used when an end-user needs to authenticate via a user agent
```
Is this not being updated to specify support for RFC 7523 [S2.2](https://www.rfc-editor.org/rfc/rfc7523#section-2.2)? I don't know if there's a use case for this in MCP, but I wanted to make sure this inconsistency was intentional and not an oversight.

D-McAdams replied:
Catching up... this comment may refer to a prior iteration. The current version uses RFC 7523 Section 2.2.
For audiences who prefer RFC 7523 Section 2.1 for workloads, that mode is being explored by the Auth Profiles working group as part of Workload Identity Federation.

maia-iyer commented:
Hello, bringing a comment from the linked issue: I'm loving this update, but want to know [why not mTLS for authentication as well](https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1046#issuecomment-3139967082)?

aaronpk commented on docs/specification/draft/basic/authorization.mdx:
```diff
@@ -166,9 +165,50 @@ these authorization servers, MCP clients will have to either:
    server).
 
-### Authorization Flow Steps
+### Resource Parameter Implementation
```
why did this section move?

D-McAdams replied:
The commit history is now long and messy. The intention is to leave the authorization.mdx file completely untouched. After futzing with git for too long, I may just create a new, clean branch to accompany the SEP for extensions.

aaronpk commented on docs/specification/draft/extensions/auth-client-credentials.mdx:
```diff
@@ -0,0 +1,161 @@
+Host: auth.example.com
+Content-Type: application/x-www-form-urlencoded
+
+grant_type=client_credentials
```
can we add `scope` to these examples as that would be best practice to request lower privileged access tokens

aaronpk commented:
This is great, thanks @D-McAdams! Left a few minor comments

D-McAdams commented:
Update:

- Incorporated prior feedback
- Created another SEP to propose MCP Extensions, a necessary prerequisite for acceptance of this change: https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1502
- Refactored this change to conform to the proposed extension mechanism
- Updated the PR title to reference both of the SEPs: MCP Extensions + OAuth Client Credential Extension

tobinsouth commented:
Just to respond to @DinoChiesa, we hear occasional feedback that a spec-level support for client credentials would be really useful.

DinoChiesa commented:
> we hear occasional feedback that a spec-level support for client credentials would be really useful.

Apparently.   But I asked _why_.  You are telling me that some people thing it would be useful.  Why? Why is it not separate?  Why is the token grant, which happens outside of the protocol and is already defined in a separate standard, part of _this protocol_?

D-McAdams commented:
Closed in favor of https://github.com/modelcontextprotocol/ext-auth/pull/3
